### PR TITLE
Use system config and user config.

### PR DESCRIPTION
**Pull request by ser-io**

### CODE REVIEW

File: pkg/cli/config.go [hunk 1]
```diff
@@ -79,3 +109,84 @@ func expandPath(path string) string {
 	}
 	return strings.ReplaceAll(path, "~", home)
 }
+
+// Build a final configuration instance from different sources. Each config parameter will take
+// the value from the first source where it's set.
+func buildConfig(sources []io.Reader) (*Config, error) {
```
Comment (jemoreira) [inline pkg/cli/config.go:115]: Would this work instead?
```
func buildConfig(sources []io.Reader) (*Config, error) {
  c := nonZeroDefaults()
  for _, s := range sources {
    if err := toml.NewDecoder(s).Decode(c); err != nil {
      return err
    }
  }
  return c, nil
}
func nonZeroDefaults() *Config {...}
```

It seems to work for JSON: https://goplay.tools/snippet/IVV86tVxlIB

**Review by jemoreira — CHANGES_REQUESTED** [review]

File: pkg/cli/config.go (hunk 1 above)
Comment (ser-io) [reply pkg/cli/config.go:94]: worked as a charm! great catch, thanks.

File: pkg/cli/config.go [hunk 2]
```diff
@@ -50,21 +52,49 @@ func (c *Config) LogFilesDeleteThreshold() time.Duration {
...
+	if _, err := os.Stat(user); err == nil {
+		userb, err = os.ReadFile(user)
+		if err != nil {
+			return nil, fmt.Errorf("Error reading user config file: %w", err)
+		}
+	}
+	if err := validateSources([]io.Reader{bytes.NewReader(userb), bytes.NewReader(defb)}); err != nil {
+		return nil, err
+	}
+	return buildConfig([]io.Reader{bytes.NewReader(userb), bytes.NewReader(defb)})
```
[the remark this answers is not in the document]
Comment (ser-io) [reply pkg/cli/config.go]: Done.

File: cmd/cvdr/main.go [hunk 3]
```diff
@@ -24,24 +24,31 @@ import (
 	"github.com/google/cloud-android-orchestration/pkg/client"
 )
 
-const configPathVar = "CVDR_CONFIG_PATH"
+const (
+	envVarConfigPathVar     = "CVDR_CONFIG_PATH"
```
Comment (jemoreira) [inline cmd/cvdr/main.go:28]: This is not necessary, use the user and default configs only and allow having any combination of them

File: build/package/host/usr/bin/cvdr [hunk 4]
```diff
@@ -14,7 +14,8 @@
 
 #!/bin/bash
 
-export CVDR_CONFIG_PATH=/etc/cvdr.toml
+export CVDR_DEFAULT_CONFIG_PATH=/etc/cvdr.toml
```
[the remark this answers is not in the document]
Comment (ser-io) [reply build/package/host/usr/bin/cvdr:3]: Done.

File: cmd/cvdr/main.go [hunk 5]
```diff
@@ -24,24 +24,31 @@ import (
...
+	dv, dok := os.LookupEnv(envVarDefaultConfigPath)
+	uv, uok := os.LookupEnv(envVarUserConfigPathVar)
+	if dok && !uok {
+		return nil, fmt.Errorf("Missing %s environment variable", envVarUserConfigPathVar)
 	}
-	defer configFile.Close()
-
-	if err := cli.ParseConfigFile(config, configFile); err != nil {
-		return fmt.Errorf("Error parsing config file: %w", err)
+	if !dok && uok {
```
[the remark this answers is not in the document]
Comment (ser-io) [reply cmd/cvdr/main.go]: Done.

File: cmd/cvdr/main.go (hunk 3 above)
Comment (ser-io) [reply cmd/cvdr/main.go]: Done.

File: pkg/cli/config.go [hunk 6]
```diff
@@ -43,33 +44,41 @@ type Config struct {
...
+func LoadWithSystemAndUserConfigFiles(system, user string) (*Config, error) {
+	systemb, err := os.ReadFile(system)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading system config file: %w", err)
+	}
+	userb, err := os.ReadFile(user)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading user config file: %w", err)
+	}
+	return buildConfig([]io.Reader{bytes.NewReader(systemb), bytes.NewReader(userb)})
```
Comment (jemoreira) [inline pkg/cli/config.go:78]: why not use the file readers in this line too?

File: cmd/cvdr/main.go [hunk 7]
```diff
@@ -24,24 +25,40 @@ import (
...
+	// User config values overrieds system config values.
+	envVarUserConfigPath = "CVDR_USER_CONFIG_PATH"
+)
 
-func readConfig(config *cli.Config) error {
-	configPath := os.Getenv(configPathVar)
-	if configPath == "" {
-		// No config file provided
-		return nil
+func loadInitialConfig() (*cli.Config, error) {
```
Comment (jemoreira) [inline cmd/cvdr/main.go:34]: I think this function could be simplified:

- Refactor `cli.LoadConfigFile` or create a new function with this signature: `func LoadConfigFile(filePath string, configObj *Config) error`
- Reimplement this function like this:
```
config := cli.DefaultConfig();
if path, ok := os.LookupEnv(envVarSystemConfigPath); ok {
  // ...
  err := cli.LoadConfigFile(path, config)
  // ...
}

if path, ok := os.LookupEnv(envVarUserConfigPath); ok {
  // ...
  // TODO if file doesn't exist generate it
  // ...
  err := cli.LoadConfig(path, config)
  // ...
}
return config
```

Comment (ser-io) [reply cmd/cvdr/main.go:34]: true, done.

File: pkg/cli/config.go (hunk 6 above)
Comment (ser-io) [reply pkg/cli/config.go:82]: true, it's no longer relevant though.

File: cmd/cvdr/main.go [hunk 8]
```diff
@@ -24,24 +25,35 @@ import (
...
+		path = cli.ExpandPath(path)
+		if err := cli.LoadConfigFile(path, config); err != nil {
+			return nil, err
+		}
 	}
-	configFile, err := os.Open(configPath)
-	if err != nil {
-		return fmt.Errorf("Error opening config file: %w", err)
+	path, ok = os.LookupEnv(envVarUserConfigPath)
+	if ok {
```
Comment (jemoreira) [inline cmd/cvdr/main.go:44]: nit: Either restrict `path` and `ok` to the scope of the `if` or use different names for system and user like you had before.

**Review by jemoreira — APPROVED** [review]

File: cmd/cvdr/main.go (hunk 8 above)
Comment (ser-io) [reply cmd/cvdr/main.go]: true, done.

**Pull request merged**